src/services/deduplicationService.ts

```diff
@@ -0,0 +1,133 @@
+import { EnteFile } from 'types/file';
+import { getEndpoint } from 'utils/common/apiUtil';
+import { getToken } from 'utils/common/key';
+import { logError } from 'utils/sentry';
+import { getLocalFiles } from './fileService';
+import HTTPService from './HTTPService';
+
+const ENDPOINT = getEndpoint();
+
+interface DuplicatesResponse {
+    duplicates: Array<{
+        fileIDs: number[];
+        size: number;
+    }>;
+}
+
+interface DuplicateFiles {
+    files: EnteFile[];
+    size: number;
+}
+
+class DeduplicationService {
+    public async getDuplicateFiles() {
+        try {
+            const dupes = await this.fetchDuplicateFileIDs();
+            let ids: number[] = [];
+            for (const dupe of dupes) {
+                ids = ids.concat(dupe.fileIDs);
+            }
+
+            const localFiles = await getLocalFiles();
+            const fileMap = new Map<number, EnteFile>();
+            for (const file of localFiles) {
+                fileMap.set(file.id, file);
+            }
+
+            const result: DuplicateFiles[] = [];
+
+            const missingFileIDs: number[] = [];
+
+            for (const dupe of dupes) {
+                const files: EnteFile[] = [];
+                for (const fileID of dupe.fileIDs) {
+                    if (fileMap.has(fileID)) {
+                        files.push(fileMap.get(fileID));
+                    } else {
+                        missingFileIDs.push(fileID);
+                    }
+                }
+
+                if (files.length > 1) {
+                    result.push({
+                        files,
+                        size: dupe.size,
+                    });
+                }
+            }
+
+            if (missingFileIDs.length > 0) {
+                logError(
+                    new Error(`Missing files: ${missingFileIDs}`),
+                    'missing files'
+                );
+            }
+
+            return result;
+        } catch (e) {
+            logError(e, 'failed to get duplicate files');
+        }
+    }
+
+    public async clubDuplicatesByTime(dupes: DuplicateFiles[]) {
```

Comment:
this is different from what I expected.  Let's say I have five files, which have the same size, but they are such that 3 of them have one timestamp while the other 2 have a different timestamp.

what this would do is show only 3 files with the same timestamp. 

but what I expected was that it would split them into two groups and show them as separate rows


did you discuss this with @vishnukvmd bhaiya ?

Reply:
I implemented the service as it was in the mobile app (https://github.com/ente-io/frame/blob/master/lib/services/deduplication_service.dart#L89), not sure why it is that way though.